### PR TITLE
Fix #11664 - has_user_defined_waypoint filter added

### LINKED ISSUE #11664: "has user-defined waypoints" filter being missed
### Tell us your idea!

A user from support misses the filter "has user-defined waypoints" (ticket 779609)

### What solution would you suggest?

_No response_

### Is there an alternative?

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by Volthai7us**
## Related issues

Fix #11664

### CODE REVIEW

**Comment by eddiemuc** [comment]
Ok to test

**Comment by Volthai7us** [comment]
I added comments to some places, I can delete them and commit again?

File: main/src/cgeo/geocaching/filters/core/StatusGeocacheFilter.java
Comment (eddiemuc) [inline main/src/cgeo/geocaching/filters/core/StatusGeocacheFilter.java]: Please remove the comments with your GitHub name (//Voltahai7us) from the PR. We don't do this (would make code pretty messy over time :-)). `git blame` can be used anytime to find out who did what if need be.

File: main/src/cgeo/geocaching/filters/core/StatusGeocacheFilter.java
Comment (eddiemuc) [inline main/src/cgeo/geocaching/filters/core/StatusGeocacheFilter.java]: Code does not follow c:geo Codacy/checkstyle rules (see breaking Codacy run), those need to be fixed. 

I recommend you install the "checkstyle" plugin in your Android studio, run it using the c:geo configuration (checkstyle.xml) and fix all findings before pushing to server. This might save you many round-trips :-)

**Review by eddiemuc — DISMISSED** [review]
Looks good apart from some "styling" stuff (see comments + red codacy check) which needs to be fixed.

**Comment by fm-sys** [comment]
@eddiemuc do you have time to invent the corresponding SQL query? Cause if we don't do it directly now, we most probably do it never...

**Comment by eddiemuc** [comment]
> @eddiemuc do you have time to invent the corresponding SQL query? Cause if we don't do it directly now, we most probably do it never...

If we feel that this is needed for a complete feature then we can't allow the PR to be merged before this is implemented as well. I am indifferent to this. Feature will work without, there's just a slight (in practice non-noteable) performance peak to it. I suggest to either create a separate issue for this or block this PR in review until we are satisfied. What do you think?

File: main/src/cgeo/geocaching/filters/core/StatusGeocacheFilter.java
Comment (eddiemuc) [reply main/src/cgeo/geocaching/filters/core/StatusGeocacheFilter.java]: ok now, thank you

File: main/src/cgeo/geocaching/filters/core/StatusGeocacheFilter.java
Comment (eddiemuc) [reply main/src/cgeo/geocaching/filters/core/StatusGeocacheFilter.java]: ok now, thank you

**Comment by Volthai7us** [comment]
After I finished addToSql method implementation should I provide new PR or commit to this PR.

**Comment by moving-bits** [comment]
> After I finished addToSql method implementation should I provide new PR or commit to this PR.

You can commit to this PR.

File: main/src/cgeo/geocaching/filters/core/StatusGeocacheFilter.java
Comment (SammysHP) [inline main/src/cgeo/geocaching/filters/core/StatusGeocacheFilter.java:232]:
```suggestion
    public Boolean getStatusHasUserdefinedWaypoints() { return statusHasUserdefinedWaypoints; }
```

File: main/src/cgeo/geocaching/filters/core/StatusGeocacheFilter.java
Comment (SammysHP) [inline main/src/cgeo/geocaching/filters/core/StatusGeocacheFilter.java:115]:
```suggestion
            (statusHasUserdefinedWaypoints == null || cache.hasUserdefinedWaypoints() == statusHasUserdefinedWaypoints) &&
```

**Review by SammysHP — CHANGES_REQUESTED** [review]
I suggest to keep the naming consistent to `cache.hasUserdefinedWaypoints()`.

**Comment by Volthai7us** [comment]
@SammysHP I will apply your suggest when I commit sql part.

**Comment by Lineflyer** [comment]
FYI:
I took the freedom to reference the issue in this PR title (which allows auto-close once merged) and set the "Work in progress" label for the time being.

**Comment by fm-sys** [comment]
> FYI: I took the freedom to reference the issue in this PR title (which allows auto-close once merged) and set the "Work in progress" label for the time being.

PR title will not work properly in all cases. The "fix #..." should be contained in the PR description or commit message. I added it to the PR description now :)

**Comment by Volthai7us** [comment]
@moving-bits could you check again?

**Review by eddiemuc — APPROVED** [review]
Looks good to me. Thanks a lot for the contribution!

I won't merge since review of @moving-bits is requested and there's still a pending change request from @SammysHP

**Review by fm-sys — COMMENTED** [review]
Seems like I've discovered two small copy-and-past mistakes. Maybe you can take a look… Apart from that it looks fine, thanks for the contribution 👍

**Comment by fm-sys** [comment]
> @fm-sys Sorry I will fixed these problem in 2 hours

All fine, don't hurry :)

**Review by moving-bits — CHANGES_REQUESTED** [review]
LGTM generally, thanks for your contribution!
Two things are left to do: As there have been other code changes meanwhile, targeting the same code parts, you'll have to do a so-called "rebase", resolving any conflicts along the way. (Lookup `git rebase upstream/master`, if you're unfamiliar with that.)
Also some variable naming inconsistencies are still present, see the remark from @SammysHP (use plurals form for `statusHasUserDefinedWaypoint` => `statusHasUserDefinedWaypoints` and `getStatusHasUserDefinedWaypoint` => `getStatusHasUserDefinedWaypoints`)

**Comment by fm-sys** [comment]
> As there have been other code changes meanwhile, targeting the same code parts, you'll have to do a so-called "rebase", resolving any conflicts along the way.

As this can be quite confusing for git beginners, I  actually would recommend not to do so.

@moving-bits
While there are conflicts when selecting `rebase merge`, it however works just fine if you select `squash and merge`. The PR should be squashed while merging anyway, so no action required from @Volthai7us IMHO. 

Only the `Waypoint` -> `Waypoints` thing is still outstanding... 

Screenshot_20220210_064407_org mozilla firefox_copy_269x151

**Comment by Volthai7us** [comment]
I hope it's okey now @fm-sys @moving-bits

**Review by moving-bits — APPROVED** [review]
> @moving-bits While there are conflicts when selecting `rebase and merge`, it however works just fine if you select `squash and merge`.

Ah, you're right, thanks for pointing this out.

> Only the `Waypoint` -> `Waypoints` thing is still outstanding...

Thanks @Volthai7us for fixing this. To my understanding all open review requests are solved by now. Have removed the "WIP" label therefore.

**Comment by moving-bits** [comment]
Let's merge here now. Thanks @Volthai7us for your contribution.

**Pull request merged**
